javascript/ql/src/Security/CWE-327/BadRandomness.ql

```diff
@@ -0,0 +1,109 @@
+/**
+ * @name Creating biased random from unbiased source.
+ * @description Some mathematical operations on random numbers can cause bias in
+ *              the results and compromise security.
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @id js/biased-cryptographic-random
+ * @tags security
+ *       external/cwe/cwe-327
+ */
+
+import javascript
+private import semmle.javascript.dataflow.InferredTypes
+
+/**
+ * Gets a source of cryptographically random numbers.
+ */
+DataFlow::SourceNode randomSource() {
+  result = DataFlow::moduleMember("crypto", ["randomBytes", "randomFillSync"]).getACall()
+  or
+  exists(DataFlow::CallNode call |
+    call = DataFlow::moduleMember("crypto", ["randomFill", "randomFillSync"]) and
+    result = call.getArgument(0).getALocalSource()
+  )
+  or
+  result = DataFlow::globalVarRef("crypto").getAMethodCall("getRandomValues")
+}
+
+/**
+ * Gets a reference to a cryptographically random number, type tracked using `t`.
+ */
+DataFlow::SourceNode random(DataFlow::TypeTracker t) {
+  t.start() and
+  result = randomSource()
+  or
+  // the random generators generally return arrays/Buffers of random numbers, we therefore track through an indexed read.
+  exists(DataFlow::PropRead read |
+    read.getBase() = random(t.continue()) and
+    exists(read.getPropertyNameExpr())
+  )
+  or
+  // reading a number from a Buffer.
+  exists(DataFlow::MethodCallNode call |
+    call.getReceiver().getALocalSource() = random(t.continue()) and
+    call.getMethodName().regexpMatch("read(BigInt|BigUInt|Double|Float|Int|UInt).{0,4}")
+  )
+  or
+  exists(DataFlow::TypeTracker t2 | result = random(t2).track(t2, t))
+}
+
+/**
+ * Gets a reference to a cryptographically random number.
+ */
+DataFlow::SourceNode random() { result = random(DataFlow::TypeTracker::end()) }
+
+/**
+ * Gets a node that that produces a biased result from otherwise cryptographically secure random numbers.
+ */
+DataFlow::Node badCrypto(string description) {
+  exists(BinaryExpr binop | result.asExpr() = binop |
+    random().flowsToExpr(binop.getLeftOperand()) and
+    random().flowsToExpr(binop.getRightOperand()) and
+    (
+      binop.getOperator() = "+" and description = "addition"
+      or
+      binop.getOperator() = "*" and description = "multiplication"
+    )
+  )
+  or
+  exists(BinaryExpr div | result.asExpr() = div and div.getOperator() = "/" |
+    random().flowsToExpr(div.getLeftOperand()) and
+    div.getRightOperand() instanceof NumberLiteral and
+    description = "division"
+  )
+  or
+  exists(BinaryExpr mod, DataFlow::SourceNode random |
+    result.asExpr() = mod and mod.getOperator() = "%"
+  |
+    random() = random and
+    random.flowsToExpr(mod.getLeftOperand()) and
+    not mod.getRightOperand().getIntValue() = [2, 4, 8, 16, 32, 64, 128] and
```

Comment:
These magic numbers also require an explanation (like the one in the test file).